### ISSUE TITLE
 [BUG] Failed to update cluster [local]: Internal error occurred: failed calling webhook "rancher.cattle.io.clusters.management.cattle.io"

### ISSUE THREAD

**Issue by mazha20** [rancher/rancher#42699]
**Rancher Server Setup**
- Rancher version: v2.7.5
- Installation option -Docker install:
   - Kubernetes Cluster version: v1.26.6+k3s1

**Information about the Cluster**
- Kubernetes version: v1.26.6+k3s1
- Cluster Type: Local

**User Information**
- What is the role of the user logged in: Admin

**Describe the bug**

**Comment by leducthinh2408** [comment]
the same

**Comment by gabrielaomoraes** [comment]
@mazha20 can you please add more details on how to reproduce the bug

**Comment by DaniDD** [comment]
Same here. Rancher is v2.7.9 an I will Update from v1.26.8 to v1.26.9. I become:
`Internal error occurred: failed calling webhook "rancher.cattle.io.clusters.management.cattle.io": failed to call webhook: Post "https://rancher-webhook.cattle-system.svc:443/v1/webhook/mutation/clusters.management.cattle.io?timeout=10s": no endpoints available for service "rancher-webhook"`

Service rancher-webhook and the Pod works fine without errors

**Comment by DirtyNative** [comment]
I also get this error since some days.. Nothing changed in my system (at least I did not do anything on my own) and out of a sudden I get this error

**Comment by DaniDD** [comment]
I think Rancher/K8s is a freaky gaming system where users have much issues, but there are no solutions :/

**Comment by mike12806** [comment]
I'm getting this error in Rancher 2.8.2

**Comment by jdbaudean** [comment]
I'm getting this error in one of our test clusters, Rancher 2.8.2 as well

**Comment by jdbaudean** [comment]
I noticed my local K3s cluster only had 2 manager nodes, but our Autoscale Group showed 3, so there was one ec2 instance that either never joined properly or was removed from the cluster at some point.  I terminated that node and the downstream RKE2 cluster almost immediately flipped back to active, and a worker that had been failing to join for some time finally joined.  Before this I manually killed the rancher-webhook pod and then redeployed Rancher itself, neither of those things helped, it must have been a bad node.

Long story short, try cycling out the manager nodes in your local cluster and see if that helps.

**Comment by daleckystepan** [comment]
Same on 2.8.5 upgrade from 2.8.4. Rancher does not come up anymore.

**Comment by KevinHuang40856** [comment]
this error will happen if you had installed rancher helm chart before and the rancher webhook pod not up running when you update cluster. because rancher has installed two webhooks, you can check them by below commands

kubectl get -n cattle-system MutatingWebhookConfiguration rancher.cattle.io
kubectl get -n cattle-system validatingwebhookconfigurations rancher.cattle.io 

in some cases, rancher webhook pod is not up running when upgrade cluster. you can run below to delete them. 

kubectl delete -n cattle-system MutatingWebhookConfiguration rancher.cattle.io
kubectl delete -n cattle-system validatingwebhookconfigurations rancher.cattle.io 

then re-run helm install/upgrade rancher-helm-chart, it will re-create these webhooks.

**Comment by huangxy29** [comment]
> validatingwebhookconfigurations

thanks very much, it works, but it is no namespace

**Comment by maituananh** [comment]
Hi in my case, my rancher version is latest. (can not update helm continue)
after running these cli, do we need to reset k8s or server ubuntu?
kubectl delete -n cattle-system MutatingWebhookConfiguration rancher.cattle.io
kubectl delete -n cattle-system validatingwebhookconfigurations rancher.cattle.io

my issue here 
https://github.com/rancher/rancher/issues/54739

**Comment by forza0310** [comment]
I encountered this error every time I started Rancher deployed via Docker. After investigation, I found that rancher-webhook starts slowly and k3s does not wait long enough for it, eventually triggering the error:
`Internal error occurred: failed calling webhook "rancher.cattle.io.clusters.management.cattle.io": failed to call webhook: Post "https://rancher-webhook.cattle-system.svc:443/v1/webhook/mutation/clusters.management.cattle.io?timeout=10s": no endpoints available for service "rancher-webhook"`

Temporary workaround – Remove the admission validation for all custom resources (e.g., authconfigs, clusters, tokens) to access the Rancher UI:
```
kubectl delete MutatingWebhookConfiguration rancher.cattle.io --ignore-not-found=true
kubectl delete ValidatingWebhookConfiguration rancher.cattle.io --ignore-not-found=true
```

Permanent fix – Following the official documentation at [Rancher Webhook Configuration](https://ranchermanager.docs.rancher.com/reference-guides/rancher-webhook#customizing-rancher-webhook-configuration), add custom Helm values when installing rancher-webhook via Helm. Rancher checks for these custom values during chart installation. The custom values must be placed in a ConfigMap named rancher-config in the cattle-system namespace, under the data key rancher-webhook.

This gives the webhook more time to initialize before k3s checks its readiness. Once this ConfigMap is created, Helm will automatically update rancher-webhook.
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: rancher-config
  namespace: cattle-system
  labels:
    app.kubernetes.io/part-of: "rancher"
data:
  rancher-webhook: '{"readinessProbe": {"initialDelaySeconds": 30, "periodSeconds": 10}}'
```

**Comment by daleckystepan** [comment]
I spent multiple hours to debug this issue and I believe I finally have something valuable.

**Error message**
```
W0525 15:38:29.918961     170 dispatcher.go:217] Failed calling webhook, failing closed rancher.cattle.io.namespaces.create-non-kubesystem: failed calling webhook "rancher.cattle.io.namespaces.create-non-kubesystem": failed to call webhook: Post "https://rancher-webhook.cattle-system.svc:443/v1/webhook/validation/namespaces?timeout=10s": no endpoints available for service "rancher-webhook"
```

**What I have noticed?**
In this state k3s does not even update Deployments, ReplicaSets, Pods, EndpointSlices, ... so basically internal controllers are not working at all. 

**How to put it together?**
The most interesting part is not updating EndpointSlices because it is directly related to the error message `no endpoints available for service`.

**Root cause**
Rancher crashes too early and internal k3s components are not able to update all resources properly. In other words: Rancher crash -> container crash -> control plane does not have time to do its job. So this is clearly the reason why Docker deployment is strongly discouraged, especially in PROD.

**Fix**
- For Rancher 2.13.3 only, for different version please find out specific k3s version
- Expects rancher data folder in `/opt/rancher`
- Do not copy&paste only and adapt command according to your environment to prevent data corruption.
```
docker run --name k3s-server --restart unless-stopped --privileged --network bridge --tmpfs /run --tmpfs /var/run --ulimit nproc=65535 --ulimit nofile=65535:65535 -v /opt/rancher/k3s:/var/lib/rancher/k3s rancher/k3s:v1.34.1-k3s1 server --node-name local-node
```
Let it run for multiple minutes. Control plane will finally get the chance to fix not updated resources. Then you can stop it and run your Rancher container.

**Proposed Solution**
A) Wait longer for k3s to be up and give it a few more minutes to let control plane run. This will make Rancher start significantly slower but reduce risk of issues.
B) Wait longer after rancher crash and do not stop k3s immediately. Probably better solution to give control plane some time. This will crashloopback slower but this is not usually problem.
C) Re-try with backoff period for webhook calls. Probably the best solution.

https://github.com/rancher/rancher/pull/55257

**Summary**
As the mentioned command help it is from my point of view clear evidence that control plane is not able to do its job because main container crashes too early and this is the limitation of Docker deployment.

**Comment by cm-patterson** [comment]
> this error will happen if you had installed rancher helm chart before and the rancher webhook pod not up running when you update cluster. because rancher has installed two webhooks, you can check them by below commands
> 
> kubectl get -n cattle-system MutatingWebhookConfiguration rancher.cattle.io kubectl get -n cattle-system validatingwebhookconfigurations rancher.cattle.io
> 
> in some cases, rancher webhook pod is not up running when upgrade cluster. you can run below to delete them.
> 
> kubectl delete -n cattle-system MutatingWebhookConfiguration rancher.cattle.io kubectl delete -n cattle-system validatingwebhookconfigurations rancher.cattle.io
> 
> then re-run helm install/upgrade rancher-helm-chart, it will re-create these webhooks.

I ran into this Chicken-Egg scenario on AWS EKS when trying to update the VPC CNI Driver.  Even though Rancher is just a "passenger" on this cluster, it annotated all the resources, preventing the managed addon from installing because rancher webhook was down. Catch-22. Thanks for this helpful hints.

**Comment by shakenfr** [comment]
hi,

after workaround

kubectl delete -n cattle-system MutatingWebhookConfiguration rancher.cattle.io
kubectl delete -n cattle-system validatingwebhookconfigurations rancher.cattle.io

It allow me to create/delete object but rancher not work anymore. It's running but when I try to use GUI
it tells me to enter password

kubectl get secret --namespace cattle-system bootstrap-secret -o go-template='{{.data.bootstrapPassword|base64decode}}{{"\n"}}'

I have this secret and when I put the password the GUI send me Invalid username or password. Please try again.
But I put the correct password ... strange

Another thing in the rancher pod logs
2026/09/01 08:52:13 [INFO] telemetry manager info not available yet, re-queing check...
2026/09/01 08:52:28 [INFO] initializing required info for telemetry manager...

but I have a metrics server up

any ideas ?
